File: oxidize-golang/core/model/prefix_cache.go
package model

import (
	"crypto/sha1"
	"encoding/hex"
	"sync"
)

// PrefixHash mirrors PrefixHash.
type PrefixHash [20]byte

// String returns the hex-encoded hash.
func (p PrefixHash) String() string { return hex.EncodeToString(p[:]) }

// ComputePrefixHash returns a stable hash for the given token prefix.
func ComputePrefixHash(prefix []Token) PrefixHash {
	sum := sha1.Sum(prefixBytes(prefix))
	var h PrefixHash
	copy(h[:], sum[:])
	return h
}

func prefixBytes(prefix []Token) []byte {
	out := make([]byte, 0, len(prefix)*4)
	for _, t := range prefix {
		out = append(out, byte(t), byte(t>>8), byte(t>>16), byte(t>>24))
	}
	return out
}

// CachedPrefix mirrors CachedPrefix.
type CachedPrefix struct {
	Hash         PrefixHash
	Tokens       []Token
	Length       int
	SharedLayers int
}

// PrefixCache mirrors PrefixCache.
type PrefixCache struct {
	mu      sync.RWMutex
	entries map[PrefixHash]*CachedPrefix
	limit   int
	hits    int
	misses  int
}

// NewPrefixCache constructs a cache with a maximum number of entries.
func NewPrefixCache(limit int) *PrefixCache {
	if limit <= 0 {
		limit = 32
	}
	return &PrefixCache{entries: map[PrefixHash]*CachedPrefix{}, limit: limit}
}

// Lookup returns the cached prefix and true if present. It mutates the
// hit/miss counters, so it must take the full write lock to avoid data
// races with concurrent callers.
func (c *PrefixCache) Lookup(prefix []Token) (*CachedPrefix, bool) {
	h := ComputePrefixHash(prefix)
	c.mu.Lock()
	defer c.mu.Unlock()
	entry, ok := c.entries[h]
	if ok {
		c.hits++
		return entry, true
	}
	c.misses++
	return nil, false
}

// Insert adds a prefix to the cache, evicting the oldest if necessary.
func (c *PrefixCache) Insert(prefix []Token, sharedLayers int) *CachedPrefix {
	h := ComputePrefixHash(prefix)
	entry := &CachedPrefix{Hash: h, Tokens: append([]Token(nil), prefix...), Length: len(prefix), SharedLayers: sharedLayers}
	c.mu.Lock()
	defer c.mu.Unlock()
	if len(c.entries) >= c.limit {
		var oldest *CachedPrefix
		for _, e := range c.entries {
			if oldest == nil || e.Length < oldest.Length {
				oldest = e
			}
		}
		if oldest != nil {
			delete(c.entries, oldest.Hash)
		}
	}
	c.entries[h] = entry
	return entry
}

// Clear removes all entries.
func (c *PrefixCache) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.entries = map[PrefixHash]*CachedPrefix{}
}

// Stats returns hits and misses.
func (c *PrefixCache) Stats() (int, int) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.hits, c.misses
}

// HitRate returns the hit ratio.
func (c *PrefixCache) HitRate() float32 {
	c.mu.RLock()
	defer c.mu.RUnlock()
	total := c.hits + c.misses
	if total == 0 {
		return 0
	}
	return float32(c.hits) / float32(total)
}

// PrefixMatcher mirrors PrefixMatcher.
type PrefixMatcher struct {
	Cache *PrefixCache
}

// NewPrefixMatcher constructs a matcher.
func NewPrefixMatcher(cache *PrefixCache) *PrefixMatcher { return &PrefixMatcher{Cache: cache} }

// LongestMatch returns the longest cached prefix for the given tokens.
func (m *PrefixMatcher) LongestMatch(tokens []Token) *CachedPrefix {
	if m.Cache == nil {
		return nil
	}
	for n := len(tokens); n > 0; n-- {
		entry, ok := m.Cache.Lookup(tokens[:n])
		if ok {
			return entry
		}
	}
	return nil
}

// CanReuseState reports whether the prefix can be reused for a new request.
func (m *PrefixMatcher) CanReuseState(prefix []Token) bool {
	entry, ok := m.Cache.Lookup(prefix)
	return ok && entry.SharedLayers > 0
}

// PrefixCacheConfig mirrors PrefixCacheConfig.
type PrefixCacheConfig struct {
	MaxEntries      int
	MinPrefixLength int
	EnableLRU       bool
}

// DefaultPrefixCacheConfig returns sensible defaults.
func DefaultPrefixCacheConfig() PrefixCacheConfig {
	return PrefixCacheConfig{MaxEntries: 64, MinPrefixLength: 4, EnableLRU: true}
}

// BuildPrefixCache constructs a cache from the config.
func BuildPrefixCache(cfg PrefixCacheConfig) *PrefixCache {
	return NewPrefixCache(cfg.MaxEntries)
}

File: oxidize-python/oxidize_python/core/model/prefix_cache.py
"""Prefix cache mirroring oxidize-golang/core/model/prefix_cache.go."""

from __future__ import annotations

import hashlib
import threading
from dataclasses import dataclass

from oxidize_python.core.model.model import Token

PrefixHash = bytes


def compute_prefix_hash(prefix: list[Token]) -> PrefixHash:
    buf = bytearray()
    for t in prefix:
        buf.extend((int(t) & 0xFFFFFFFF).to_bytes(4, "little", signed=False))
    return hashlib.sha1(buf).digest()


@dataclass
class CachedPrefix:
    hash: PrefixHash
    tokens: list[Token]
    length: int
    shared_layers: int


class PrefixCache:
    def __init__(self, limit: int = 32) -> None:
        self._limit = limit if limit > 0 else 32
        self._entries: dict[PrefixHash, CachedPrefix] = {}
        self._mu = threading.Lock()
        self._hits = 0
        self._misses = 0

    def lookup(self, prefix: list[Token]) -> tuple[CachedPrefix | None, bool]:
        h = compute_prefix_hash(prefix)
        with self._mu:
            entry = self._entries.get(h)
            if entry is not None:
                self._hits += 1
                return entry, True
            self._misses += 1
            return None, False

    def insert(self, prefix: list[Token], shared_layers: int) -> CachedPrefix:
        h = compute_prefix_hash(prefix)
        entry = CachedPrefix(
            hash=h,
            tokens=list(prefix),
            length=len(prefix),
            shared_layers=shared_layers,
        )
        with self._mu:
            if len(self._entries) >= self._limit:
                oldest: CachedPrefix | None = None
                for e in self._entries.values():
                    if oldest is None or e.length < oldest.length:
                        oldest = e
                if oldest is not None:
                    del self._entries[oldest.hash]
            self._entries[h] = entry
        return entry

    def clear(self) -> None:
        with self._mu:
            self._entries.clear()

    def stats(self) -> tuple[int, int]:
        with self._mu:
            return self._hits, self._misses

    def hit_rate(self) -> float:
        with self._mu:
            total = self._hits + self._misses
            if total == 0:
                return 0.0
            return self._hits / total


class PrefixMatcher:
    def __init__(self, cache: PrefixCache | None) -> None:
        self.cache = cache

    def longest_match(self, tokens: list[Token]) -> CachedPrefix | None:
        if self.cache is None:
            return None
        for n in range(len(tokens), 0, -1):
            entry, ok = self.cache.lookup(tokens[:n])
            if ok:
                return entry
        return None

    def can_reuse_state(self, prefix: list[Token]) -> bool:
        entry, ok = self.cache.lookup(prefix) if self.cache else (None, False)
        return ok and entry is not None and entry.shared_layers > 0


@dataclass
class PrefixCacheConfig:
    max_entries: int = 64
    min_prefix_length: int = 4
    enable_lru: bool = True


def default_prefix_cache_config() -> PrefixCacheConfig:
    return PrefixCacheConfig()


def build_prefix_cache(cfg: PrefixCacheConfig) -> PrefixCache:
    return PrefixCache(cfg.max_entries)
